File: js/AdsNetworkWrapper.js
var adsWrapper =
    {
        $adsLocal: {
            onMraidAdsSdkReady : function (w4Self, callback)
            {
                /** MRAID **/

                if (typeof mraid !== "undefined")
                {
                    mraid.addEventListener
                    (
                        'viewableChange',
                        function (viewable)
                        {
                            if (viewable) { wasmTable.get(callback)(w4Self, true); }
                            else {  wasmTable.get(callback)(w4Self, false); }
                        }
                    );
                }
            },

            onIronsourceAdsSdkReady : function (w4Self, callbackviewableChange, callbackAudioVolumeChanged)
            {
                /** Ironsource **/
                if (typeof  dapi  != "undefined")
                {
                    //dapi.removeEventListener("ready", adsLocal.onIronsourceAdsSdkReady);

                    dapi.getScreenSize();
                    dapi.getAudioVolume();

                    if (typeof callbackAudioVolumeChanged !== "undefined")
                    {
                        dapi.addEventListener(
                            "audioVolumeChange",
                            function (event) {
                                wasmTable.get(callbackAudioVolumeChanged)(w4Self, event);
                            }
                        );
                    }

                    if (typeof callbackviewableChange !== "undefined")
                    {
                        dapi.addEventListener(
                            "viewableChange",
                            function (event) {
                                if (event.isViewable) {
                                    wasmTable.get(callbackviewableChange)(w4Self, true);
                                } else {
                                    wasmTable.get(callbackviewableChange)(w4Self, false);
                                }
                            }
                        );
                    }

                }
            }
        },

        w4AdsStart : function ()
        {
            /** MRAID **/
            if (typeof mraid !== "undefined")
            {
                if (typeof TARGET_URL === "undefined")
                {
                    console.error("TARGET_URL is undef, set https://www.space.com/");
                    TARGET_URL = "https://www.space.com/";
                }
            }
        },

        w4AdsComplete : function ()
        {
            /** Vungle **/
            if (typeof parent !== "undefined" && parent != null)
            {
                parent.postMessage('complete','*');
            }

        },

        w4AdsClose : function ()
        {
            /** MRAID **/
            if (typeof mraid !== "undefined")
            {
                mraid.close();
            }
        },

        w4AdsUseCustomCloseButton : function ( flag )
        {
            /** MRAID **/
            if (typeof mraid !== "undefined")
            {
                mraid.useCustomClose(flag);
            }
        },

        w4AdsIsCustomButtonAvalible : function (  )
        {
            /** MRAID **/
            if (typeof mraid !== "undefined")
            {
                return true;
            }

            return false;
        },

        w4AdsCallCTA:function ( )
        {
            /** MRAID **/
            if (typeof mraid !== "undefined")
            {
                mraid.open(TARGET_URL);
            }

            /** Google Ads**/
            if (typeof ExitApi !==  "undefined")
            {
                ExitApi.exit();
            }

            /** Facebook **/
            if (typeof FbPlayableAd !== "undefined")
            {
                FbPlayableAd.onCTAClick();
            }

            /** Vungle **/
            if (typeof parent !== "undefined" && parent != null)
            {
                parent.postMessage('download','*');
            }

            /** Ironsource **/
            if (typeof dapi !== "undefined")
            {
                dapi.openStoreUrl();
            }

            /** LiftOff **/
            if (typeof Liftoff !== "undefined")
            {
                Liftoff.open();
            }
        },

        w4AdsAudioVolumeChangeHandler: function (w4Self, callback)
        {
            /** Ironsource **/
            if (typeof dapi !== "undefined")
            {
                if (!dapi.isReady())
                {
                    console.log("!dapi.isReady()");
                    dapi.addEventListener("ready", function () { adsLocal.onIronsourceAdsSdkReady(w4Self, null, callback);});
                }
                else
                {
                    console.log("dapi.isReady()");
                    adsLocal.onIronsourceAdsSdkReady(w4Self, null, callback);
                }
            }
        },
        w4AdsVisibleChangeHandler: function (w4Self, callback)
        {
            /** MRAID **/
            if (typeof mraid !== "undefined")
            {
                if (mraid.getState() === 'loading')
                {
                    mraid.addEventListener('ready', function () { adsLocal.onMraidAdsSdkReady (w4Self, callback);});
                }
                else
                {
                    adsLocal.onMraidAdsSdkReady(w4Self, callback);
                }
            }

            /** Vungle **/
            if (typeof parent !== "undefined" && parent != null)
            {
                window.addEventListener('ad-event-pause', function () { wasmTable.get(callback)(w4Self, false); });
                window.addEventListener('ad-event-resume', function () { wasmTable.get(callback)(w4Self, true); });
            }

            /** Ironsource **/
            if (typeof dapi !== "undefined")
            {
                if (!dapi.isReady())
                {
                    console.log("!dapi.isReady()");
                    dapi.addEventListener("ready", function () { adsLocal.onIronsourceAdsSdkReady(w4Self, callback, null);});
                }
                else
                {
                    console.log("dapi.isReady()");
                    adsLocal.onIronsourceAdsSdkReady(w4Self, callback, null);
                }
            }

        }
    };

autoAddDeps(adsWrapper, '$adsLocal');
mergeInto(LibraryManager.library, adsWrapper);

File: sources/AdsNetworkWrapper.cpp
#include "interface/AdsNetworkWrapper.h"

//using namespace w4::network::ads;

#ifndef __EMSCRIPTEN__
    #include <stdexcept>
#endif

#ifdef __EMSCRIPTEN__
W4_JS_IMPORT
{
    void w4AdsStart();
    void w4AdsComplete();
    void w4AdsClose();
    void w4AdsVisibleChangeHandler(void * w4Self, w4::network::ads::BoolCallback callback);
    void w4AdsAudioVolumeChangeHandler(void * w4Self, w4::network::ads::FloatCallback callback);
    void w4AdsUseCustomCloseButton(bool flag);
    bool w4AdsIsCustomButtonAvalible();
    void w4AdsCallCTA();
}

void w4::network::ads::start() { w4AdsStart(); };
void w4::network::ads::complete() { w4AdsComplete(); }
void w4::network::ads::close() { w4AdsClose(); };
void w4::network::ads::CTA() { w4AdsCallCTA(); };
void w4::network::ads::onVisibleChangeHandler(void * w4Self, w4::network::ads::BoolCallback callback) { w4AdsVisibleChangeHandler(w4Self,callback); }
void w4::network::ads::onAudioVolumeChangeHandler(void *w4Self, FloatCallback callback) { w4AdsAudioVolumeChangeHandler(w4Self,callback);}
void w4::network::ads::useCustomCloseButton(bool flag) { w4AdsUseCustomCloseButton(flag); }
bool w4::network::ads::isCustomCloseButtonAvalible() { return w4AdsIsCustomButtonAvalible();};

#else

void w4::network::ads::start() {};
void w4::network::ads::complete() {};
void w4::network::ads::close() {};
void w4::network::ads::CTA() {};
void w4::network::ads::onVisibleChangeHandler(void * w4Self, w4::network::ads::BoolCallback callback){};
void w4::network::ads::onAudioVolumeChangeHandler(void *w4Self, FloatCallback callback) { };
void w4::network::ads::useCustomCloseButton(bool flag) {};
bool w4::network::ads::isCustomCloseButtonAvalible() { return false;};

#endif